### PR TITLE
feat: Add entity storage connector for GCP Firestore

### PR DESCRIPTION
**Pull request by psyhedeliq**

### CODE REVIEW

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+		}
+	}
+
+	/**
+	 * Handle undefined values in the entity.
+	 * @param entity The entity to handle undefined values in.
+	 * @returns The entity with undefined values handled.
+	 * @internal
+	 */
+	protected handleUndefinedValues(entity: unknown): unknown {
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:360]: Probably won't need this method if we remove the option

**Review by obany — COMMENTED** [review]
First review done, a few changes, mostly will simplify the code.

The signatures for the `get`, `set` and `remove` methods have also been updated in the base connector to have optional equality conditions, so the code will need to be updated to sort this. The dynamoDb and Scylla connectors are already updated.

File: packages/entity-storage-connector-gcp-firestore/tests/.env [hunk 2]
```diff
@@ -1,9 +1,7 @@
 TEST_FIRESTORE_PROJECT_ID=your-project-id
 TEST_FIRESTORE_COLLECTION_NAME=your-collection-name
-TEST_FIRESTORE_CLIENT_EMAIL=your-client-email
-TEST_FIRESTORE_PRIVATE_KEY=your-private-key
-TEST_FIRESTORE_UNDEFINED_VALUE_HANDLING=your-preferred-handling
 TEST_FIRESTORE_KEY_FILENAME=path-to-your-key-file
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/tests/.env:3]: KEY_FILENAME can be removed as well

File: packages/entity-storage-connector-gcp-firestore/tests/setupTestEnv.ts [hunk 3]
```diff
@@ -16,27 +15,12 @@ Guards.stringValue(
...
-	"TEST_FIRESTORE_CLIENT_EMAIL",
-	process.env.TEST_FIRESTORE_CLIENT_EMAIL
-);
-Guards.stringValue("TestEnv", "TEST_FIRESTORE_PRIVATE_KEY", process.env.TEST_FIRESTORE_PRIVATE_KEY);
-Guards.stringValue(
-	"TestEnv",
-	"TEST_FIRESTORE_UNDEFINED_VALUE_HANDLING",
-	process.env.TEST_FIRESTORE_UNDEFINED_VALUE_HANDLING
-);
 Guards.stringValue(
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/tests/setupTestEnv.ts:18]: remove key_filename

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+	public getSchema(): IEntitySchema {
+		return this._entitySchema as IEntitySchema;
+	}
+
+	/**
+	 * Get an entity.
+	 * @param id The id of the entity to get.
+	 * @returns The object if it can be found or undefined.
+	 */
+	public async get(id: string): Promise<T | undefined> {
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:187]: Update with new signature which includes `conditions`

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+			throw new GeneralError(this.CLASS_NAME, "getEntityFailed", { id }, err);
+		}
+	}
+
+	/**
+	 * Set an entity.
+	 * @param entity The entity to set.
+	 * @returns Nothing.
+	 */
+	public async set(entity: T): Promise<void> {
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:207]: Update with new signature which includes `conditions`

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+			throw new GeneralError(this.CLASS_NAME, "setEntityFailed", { entity }, err);
+		}
+	}
+
+	/**
+	 * Remove the entity.
+	 * @param id The id of the entity to remove.
+	 * @returns Nothing.
+	 */
+	public async remove(id: string): Promise<void> {
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:239]: Update with new signature which includes `conditions`

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 7]
```diff
@@ -182,17 +180,46 @@ export class FirestoreEntityStorageConnector<T = unknown> implements IEntityStor
...
+		id: string,
+		secondaryIndex?: keyof T,
+		conditions?: { property: keyof T; value: unknown }[]
+	): Promise<T | undefined> {
 		Guards.stringValue(this.CLASS_NAME, nameof(id), id);
 
 		try {
-			const docRef = this.getCollection().doc(id);
-			const doc = await docRef.get();
+			if (!secondaryIndex && (!conditions || conditions.length === 0)) {
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:195]: You can use the Is methods to make this stricter
`if (!Is.StringValue(secondaryIndex) && !Is.arrayValue(conditions))`

File: packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts [hunk 8]
```diff
@@ -202,46 +229,107 @@ export class FirestoreEntityStorageConnector<T = unknown> implements IEntityStor
...
+			} else {
+				await this._firestoreClient.runTransaction(async transaction => {
+					const docSnapshot = await transaction.get(docRef);
+
+					if (!docSnapshot.exists) {
+						transaction.set(docRef, entityCopy as DocumentData);
+					} else {
+						const data = docSnapshot.data() as T;
+
+						let conditionsMet = true;
```
Comment (obany) [inline packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:264]: I am guessing you couldnt find a way to do this natively without get, check and then set

Comment (psyhedeliq) [reply packages/entity-storage-connector-gcp-firestore/src/firestoreEntityStorageConnector.ts:264]: Yes, you're correct. Due to Firestore's limitations, we need to perform a get, check, and then set within a transaction to ensure that the update occurs only if the specified conditions are met. Firestore's native preconditions don't support conditional updates based on arbitrary field values without reading the document first.

**Review by obany — APPROVED** [review]

**Pull request merged**